File: 프로그래머스/동적계획법/N으로표현.java
package 프로그래머스.동적계획법;

import java.util.*;

public class N으로표현 {
    public static void main(String args[]){
        int N = 5;
        int number = 12;
        System.out.println(solution(N, number));

    }

    public static int solution(int N, int number) {
        int answer = 0;
        ArrayList<Set<Integer>> dp = new ArrayList<>();
        for(int i=0;i<9;i++){
            dp.add(new HashSet<>());
        }
        dp.get(1).add(N);
        System.out.println(dp.get(1));

        for(int i=2;i<9;i++){
            Set<Integer> countSet = dp.get(i);
            for(int j=1;j<i;j++){
                Set<Integer> preSet = dp.get(j);
                Set<Integer> postSet = dp.get(i-j);

                for(int preNum : preSet){
                    for(int postNum : postSet){
                        countSet.add(preNum+postNum);
                        countSet.add(preNum-postNum);
                        countSet.add(preNum*postNum);
                        if(preNum!=0 && postNum!=0){
                            countSet.add(preNum/postNum);
                        }
                    }
                }
            }
            countSet.add(Integer.parseInt(String.valueOf(N).repeat(i)));

            System.out.println(countSet);
            if(countSet.contains(number)){
                answer = i;
                break;
            }
        }

        if(answer==0) answer=-1;

        return answer;

    }
    
}

File: 프로그래머스/동적계획법/N으로표현.py
def solution(N, number):
    
    dp = []
    for i in range(1, 9):
        tmp_set = set()
        tmp_set.add(int(str(N)*i))
        for j in range(i-1):
            print(f'i: {i}, j:{j}, i-j:{i-j}')
            for prevNum in dp[j]:
                for postNum in dp[-j-1]:
                    tmp_set.add(prevNum+postNum)
                    tmp_set.add(prevNum-postNum)
                    tmp_set.add(prevNum*postNum)
                    if postNum!=0 : tmp_set.add(prevNum/postNum)
                
        if (number in tmp_set)==True:
            return i
        
        print(tmp_set)
        dp.append(tmp_set)
    
    return -1

N = 5
number = 12

print(solution(N, number))